src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,587 @@
+    /// Logs may include a StartTime. If present, it represents seconds since the epoch.
+    StartTime(f64),
+    /// Logs may include a BaseTime. If present, it represents seconds since the epoch.
+    BaseTime(f64),
+    /// An individual interval histogram.
+    Interval(IntervalLogHistogram<'a>),
+}
+
+/// Errors that occur when parsing an interval log.
+#[derive(Debug, PartialEq)]
+pub enum LogIteratorError {
+    /// Parsing failed
+    ParseError {
+        /// Offset in the input where the failed parse started
+        offset: usize,
+    },
+}
+
+/// Parse interval logs.
+///
+/// This iterator exposes each item (excluding comments and other information-free lines). See
+/// `LogEntry`.
+///
+/// Because histogram deserialization is deferred, parsing logs is fast. See the `interval_log`
+/// benchmark if you wish to see how it does on your hardware. As a baseline, parsing a log of 1000
+/// random histograms of 10,000 values each takes 8ms total on an E5-1650v3.
+///
+/// Deferring deserialization is handy because it allows you to cheaply navigate the log to find
+/// the records you care about (e.g. ones in a certain time range, or with a certain tag) without
+/// doing all the allocation, etc, of deserialization.
+///
+/// If you're looking for a direct port of the Java impl's `HistogramLogReader`, this isn't one: it
+/// won't deserialize for you, and it pushes the burden of figuring out what to do with StartTime,
+/// BaseTime, etc to you, and there aren't built in functions to filter by timestamp. On the other
+/// hand, because it doesn't do those things, it is much more flexible: you can easily build any
+/// sort of filtering you want, not just timestamp ranges, because you have cheap access to all the
+/// metadata before incurring the cost of deserialization. If you're not using any timestamp
+/// headers, or at least using them in straightforward ways, it is easy to accumulate the
+/// timestamp state you need. Since all the parsing is taken care of already, writing your own
+/// `HistogramLogReader` equivalent that fits the way your logs are assembled is just a couple of
+/// lines. (And if you're doing complex stuff, we probably wouldn't have built something that fits
+/// your quirky logs anyway!)
+///
+/// This parses from a slice representing the complete file because it made implementation easier
+/// (and also supports mmap'd files for maximum parsing speed). If parsing from a `Read` is
+/// important for your use case, open an issue about it.
+pub struct IntervalLogIterator<'a> {
+    orig_len: usize,
+    input: &'a [u8],
+}
+
+impl<'a> IntervalLogIterator<'a> {
+    /// Create a new iterator from the UTF-8 bytes of an interval log.
+    pub fn new(input: &'a [u8]) -> IntervalLogIterator<'a> {
+        IntervalLogIterator {
+            orig_len: input.len(),
+            input,
+        }
+    }
+}
+
+impl<'a> Iterator for IntervalLogIterator<'a> {
+    type Item = Result<LogEntry<'a>, LogIteratorError>;
```

Comment:
I have some issues with this.
If I've understood the log format correctly, this iterator will *always* first yield at most one each of `StartTime` and `BaseTime`, and from that point forwards, only yield `LogEntry::Interval`. If that interpretation is correct, then it seems like this should really be an iterator that yields `IntervalLogHistogram<'a>`, with some kind of initial step that gives you the start/base time. That'd be much more convenient to work with, as it would obviate the need for that pesky `match`.

Reply:
Hmm, I suppose... It seems unfortunate, because it makes the API a fair amount less ergonomic. I guess we could provide a refined iterator on top of this, whereas the reverse is not true, so leave it like this for now.